pandas/core/dtypes/cast.py

```diff
@@ -1505,12 +1506,15 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
 
     """
     if dtype is None:
-        dtype, fill_value = infer_dtype_from_scalar(value)
+        dtype, fill_value = infer_dtype_from_scalar(value, pandas_dtype=True)
     else:
         fill_value = value
 
-    values = np.empty(shape, dtype=dtype)
-    values.fill(fill_value)
+    if type(dtype) in registry.dtypes:
+        values = dtype.construct_array_type()._from_sequence([value] * shape)
```

Comment:
What's the type on shape here? A tuple or integer?

Extension arrays are only allowed to be 1D. We'll need to make sure that the caller is able to handle this. May need to restructure the logic a bit to make sure it's not surprised.

Reply:
Yeah, this is the part that was most confusing to me.

Correct me if I'm wrong, but I believe that the motivation for this function accepting a tuple instead of an int is that similar data type columns can be grouped together in memory by creating them all within the same numpy array.

If that's correct, and if Extension arrays can only be 1D, then the same logic would not really apply to extension types, right? There would be no reason to call this function using `shape` as a tuple, since no grouping could happen anyway.

To be explicit, what I'm suggesting is perhaps only allowing `shape` to be an int when `value` is an extension type, otherwise we throw an error. We could also build in support if `shape` is a tuple that looks like `(n, 1)`, since that is virtually the same thing.

How does this sound?

Reply:
Alternatively, instead of throwing an error, it could raise a warning and set the column's type to object.

It looks like this issue is causing a test failure from this code:
```
data1 = pd.to_datetime("20100101 01:01").tz_localize("UTC")
    df1 = pd.DataFrame(
        columns=["UTCdatetime", "abc"],
    data=data1,
    index=pd.date_range("20140627", periods=1),
 )
```

@TomAugspurger Do you have any guidance on how I should handle this?

Reply:
Hmm the example you posted is a bit unfortunate...  And the fact that it's called this late into DataFrame's constructor makes this really hard. That's already complicated enough.

I would suggest a bit larger refactor if it's OK. I really don't like the idea of `cast_scalar_to_array(shape, ...)` returning anything that *doesn't* have a shape of `shape`.

So, we might have a `cast_scalar_to_extension_arrays(n_rows, scalar, dtype=None)`. And then places that currently use `cast_scalar_to_array` will check the type of the scalar. If it's an extension scalar then we create a list or dict of extension arrays by calling `cast_scalar_to_extension_arrays`.

We'll need to update that test. I'd consider it a bug that the dtype of `Series(data1, index=...)` doesn't match the dtypes of `DataFrame(data1, index=..., columns=...)`.

LMK if any of that is unclear.

Reply:
Thanks @TomAugspurger for your help, that does make sense. I updated the code per these suggestions and everything seems to be working as expected, except the `test_combine_first_timezone` test.

I've gotten pretty lost in debugging this one, I'm not sure why this test should have changed, can you provide some insight here?

Reply:
Mmm this is messy, sorry. But the short version is that the old test was relying on the buggy behavior of `DataFrame(timestamp_scalar)` returning object dtype. Notice the dtypes in the tests

```python
(Pdb) df1.dtypes
UTCdatetime    object
abc            object
dtype: object
(Pdb) df2.dtypes
UTCdatetime    object
xyz            object
dtype: object
```

On this PR those are (correctly) 

```python
(Pdb) df1.dtypes
UTCdatetime    datetime64[ns, UTC]
abc            datetime64[ns, UTC]
dtype: object
(Pdb) df2.dtypes
UTCdatetime    datetime64[ns, UTC]
xyz            datetime64[ns, UTC]
dtype: object
```

Now in `combine` you get into some messy dtype handling issues. In order to not get hung up on those, I'd recommend that you

1. update the tests in `test_combine_first_timezone` to `.astype(object)` to match the old behavior
2. Open an issue demonstrating the buggy behavior (search to make sure we don't have one yet).
3. Add a comment referencing that issue.

Reply:
Thanks for the info, that worked!